File: v2/pkg/runtime/notifications.go
package runtime

import (
	"context"

	"github.com/wailsapp/wails/v2/internal/frontend"
)

// NotificationOptions contains configuration for a notification.
type NotificationOptions = frontend.NotificationOptions

// NotificationAction represents an action button for a notification.
type NotificationAction = frontend.NotificationAction

// NotificationCategory groups actions for notifications.
type NotificationCategory = frontend.NotificationCategory

// NotificationResponse represents the response sent by interacting with a notification.
type NotificationResponse = frontend.NotificationResponse

// NotificationResult represents the result of a notification response,
// returning the response or any errors that occurred.
type NotificationResult = frontend.NotificationResult

// InitializeNotifications initializes the notification service for the application.
// This must be called before sending any notifications. On macOS, this also ensures
// the notification delegate is properly initialized.
func InitializeNotifications(ctx context.Context) error {
	fe := getFrontend(ctx)
	return fe.InitializeNotifications()
}

// CleanupNotifications cleans up notification resources and releases any held connections.
// This should be called when shutting down the application to properly release resources
// (primarily needed on Linux to close D-Bus connections).
func CleanupNotifications(ctx context.Context) {
	fe := getFrontend(ctx)
	fe.CleanupNotifications()
}

// IsNotificationAvailable checks if notifications are available on the current platform.
func IsNotificationAvailable(ctx context.Context) bool {
	fe := getFrontend(ctx)
	return fe.IsNotificationAvailable()
}

// RequestNotificationAuthorization requests notification authorization from the user.
// On macOS, this prompts the user to allow notifications. On other platforms, this
// always returns true. Returns true if authorization was granted, false otherwise.
func RequestNotificationAuthorization(ctx context.Context) (bool, error) {
	fe := getFrontend(ctx)
	return fe.RequestNotificationAuthorization()
}

// CheckNotificationAuthorization checks the current notification authorization status.
// On macOS, this checks if the app has notification permissions. On other platforms,
// this always returns true.
func CheckNotificationAuthorization(ctx context.Context) (bool, error) {
	fe := getFrontend(ctx)
	return fe.CheckNotificationAuthorization()
}

// SendNotification sends a basic notification with the given options.
// The notification will display with the provided title, subtitle (if supported),
// and body text.
func SendNotification(ctx context.Context, options NotificationOptions) error {
	fe := getFrontend(ctx)
	return fe.SendNotification(options)
}

// SendNotificationWithActions sends a notification with action buttons.
// A NotificationCategory must be registered first using RegisterNotificationCategory.
// The options.CategoryID must match a previously registered category ID.
// If the category is not found, a basic notification will be sent instead.
func SendNotificationWithActions(ctx context.Context, options NotificationOptions) error {
	fe := getFrontend(ctx)
	return fe.SendNotificationWithActions(options)
}

// RegisterNotificationCategory registers a notification category that can be used
// with SendNotificationWithActions. Categories define the action buttons and optional
// reply fields that will appear on notifications.
func RegisterNotificationCategory(ctx context.Context, category NotificationCategory) error {
	fe := getFrontend(ctx)
	return fe.RegisterNotificationCategory(category)
}

// RemoveNotificationCategory removes a previously registered notification category.
func RemoveNotificationCategory(ctx context.Context, categoryId string) error {
	fe := getFrontend(ctx)
	return fe.RemoveNotificationCategory(categoryId)
}

// RemoveAllPendingNotifications removes all pending notifications from the notification center.
// On Windows, this is a no-op as the platform manages notification lifecycle automatically.
func RemoveAllPendingNotifications(ctx context.Context) error {
	fe := getFrontend(ctx)
	return fe.RemoveAllPendingNotifications()
}

// RemovePendingNotification removes a specific pending notification by its identifier.
// On Windows, this is a no-op as the platform manages notification lifecycle automatically.
func RemovePendingNotification(ctx context.Context, identifier string) error {
	fe := getFrontend(ctx)
	return fe.RemovePendingNotification(identifier)
}

// RemoveAllDeliveredNotifications removes all delivered notifications from the notification center.
// On Windows, this is a no-op as the platform manages notification lifecycle automatically.
func RemoveAllDeliveredNotifications(ctx context.Context) error {
	fe := getFrontend(ctx)
	return fe.RemoveAllDeliveredNotifications()
}

// RemoveDeliveredNotification removes a specific delivered notification by its identifier.
// On Windows, this is a no-op as the platform manages notification lifecycle automatically.
func RemoveDeliveredNotification(ctx context.Context, identifier string) error {
	fe := getFrontend(ctx)
	return fe.RemoveDeliveredNotification(identifier)
}

// RemoveNotification removes a notification by its identifier.
// This is a convenience function that works across platforms. On macOS, use the
// more specific RemovePendingNotification or RemoveDeliveredNotification functions.
func RemoveNotification(ctx context.Context, identifier string) error {
	fe := getFrontend(ctx)
	return fe.RemoveNotification(identifier)
}

// OnNotificationResponse registers a callback function that will be invoked when
// a user interacts with a notification (e.g., clicks an action button or the notification itself).
// The callback receives a NotificationResult containing the response details or any errors.
func OnNotificationResponse(ctx context.Context, callback func(result NotificationResult)) {
	fe := getFrontend(ctx)
	fe.OnNotificationResponse(callback)
}

File: v2/internal/frontend/runtime/desktop/notifications.js
/*
 _       __      _ __
| |     / /___ _(_) /____
| | /| / / __ `/ / / ___/
| |/ |/ / /_/ / / (__  )
|__/|__/\__,_/_/_/____/
The electron alternative for Go
(c) Lea Anthony 2019-present
*/
/* jshint esversion: 9 */

import {Call} from "./calls";

/**
 * Initialize the notification service for the application.
 * This must be called before sending any notifications.
 * On macOS, this also ensures the notification delegate is properly initialized.
 *
 * @export
 * @return {Promise<void>}
 */
export function InitializeNotifications() {
    return Call(":wails:InitializeNotifications");
}

/**
 * Clean up notification resources and release any held connections.
 * This should be called when shutting down the application to properly release resources
 * (primarily needed on Linux to close D-Bus connections).
 *
 * @export
 * @return {Promise<void>}
 */
export function CleanupNotifications() {
    return Call(":wails:CleanupNotifications");
}

/**
 * Check if notifications are available on the current platform.
 *
 * @export
 * @return {Promise<boolean>} True if notifications are available, false otherwise
 */
export function IsNotificationAvailable() {
    return Call(":wails:IsNotificationAvailable");
}

/**
 * Request notification authorization from the user.
 * On macOS, this prompts the user to allow notifications.
 * On other platforms, this always returns true.
 *
 * @export
 * @return {Promise<boolean>} True if authorization was granted, false otherwise
 */
export function RequestNotificationAuthorization() {
    return Call(":wails:RequestNotificationAuthorization");
}

/**
 * Check the current notification authorization status.
 * On macOS, this checks if the app has notification permissions.
 * On other platforms, this always returns true.
 *
 * @export
 * @return {Promise<boolean>} True if authorized, false otherwise
 */
export function CheckNotificationAuthorization() {
    return Call(":wails:CheckNotificationAuthorization");
}

/**
 * Send a basic notification with the given options.
 * The notification will display with the provided title, subtitle (if supported), and body text.
 *
 * @export
 * @param {Object} options - Notification options
 * @param {string} options.id - Unique identifier for the notification
 * @param {string} options.title - Notification title
 * @param {string} [options.subtitle] - Notification subtitle (macOS and Linux only)
 * @param {string} [options.body] - Notification body text
 * @param {string} [options.categoryId] - Category ID for action buttons (requires SendNotificationWithActions)
 * @param {Object<string, any>} [options.data] - Additional user data to attach to the notification
 * @return {Promise<void>}
 */
export function SendNotification(options) {
    return Call(":wails:SendNotification", [options]);
}

/**
 * Send a notification with action buttons.
 * A NotificationCategory must be registered first using RegisterNotificationCategory.
 * The options.categoryId must match a previously registered category ID.
 * If the category is not found, a basic notification will be sent instead.
 *
 * @export
 * @param {Object} options - Notification options
 * @param {string} options.id - Unique identifier for the notification
 * @param {string} options.title - Notification title
 * @param {string} [options.subtitle] - Notification subtitle (macOS and Linux only)
 * @param {string} [options.body] - Notification body text
 * @param {string} options.categoryId - Category ID that matches a registered category
 * @param {Object<string, any>} [options.data] - Additional user data to attach to the notification
 * @return {Promise<void>}
 */
export function SendNotificationWithActions(options) {
    return Call(":wails:SendNotificationWithActions", [options]);
}

/**
 * Register a notification category that can be used with SendNotificationWithActions.
 * Categories define the action buttons and optional reply fields that will appear on notifications.
 * Registering a category with the same ID as a previously registered category will override it.
 *
 * @export
 * @param {Object} category - Notification category definition
 * @param {string} category.id - Unique identifier for the category
 * @param {Array<Object>} [category.actions] - Array of action buttons
 * @param {string} category.actions[].id - Unique identifier for the action
 * @param {string} category.actions[].title - Display title for the action button
 * @param {boolean} [category.actions[].destructive] - Whether the action is destructive (macOS-specific)
 * @param {boolean} [category.hasReplyField] - Whether to include a text input field for replies
 * @param {string} [category.replyPlaceholder] - Placeholder text for the reply field (required if hasReplyField is true)
 * @param {string} [category.replyButtonTitle] - Title for the reply button (required if hasReplyField is true)
 * @return {Promise<void>}
 */
export function RegisterNotificationCategory(category) {
    return Call(":wails:RegisterNotificationCategory", [category]);
}

/**
 * Remove a previously registered notification category.
 *
 * @export
 * @param {string} categoryId - The ID of the category to remove
 * @return {Promise<void>}
 */
export function RemoveNotificationCategory(categoryId) {
    return Call(":wails:RemoveNotificationCategory", [categoryId]);
}

/**
 * Remove all pending notifications from the notification center.
 * On Windows, this is a no-op as the platform manages notification lifecycle automatically.
 *
 * @export
 * @return {Promise<void>}
 */
export function RemoveAllPendingNotifications() {
    return Call(":wails:RemoveAllPendingNotifications");
}

/**
 * Remove a specific pending notification by its identifier.
 * On Windows, this is a no-op as the platform manages notification lifecycle automatically.
 *
 * @export
 * @param {string} identifier - The ID of the notification to remove
 * @return {Promise<void>}
 */
export function RemovePendingNotification(identifier) {
    return Call(":wails:RemovePendingNotification", [identifier]);
}

/**
 * Remove all delivered notifications from the notification center.
 * On Windows, this is a no-op as the platform manages notification lifecycle automatically.
 *
 * @export
 * @return {Promise<void>}
 */
export function RemoveAllDeliveredNotifications() {
    return Call(":wails:RemoveAllDeliveredNotifications");
}

/**
 * Remove a specific delivered notification by its identifier.
 * On Windows, this is a no-op as the platform manages notification lifecycle automatically.
 *
 * @export
 * @param {string} identifier - The ID of the notification to remove
 * @return {Promise<void>}
 */
export function RemoveDeliveredNotification(identifier) {
    return Call(":wails:RemoveDeliveredNotification", [identifier]);
}

/**
 * Remove a notification by its identifier.
 * This is a convenience function that works across platforms.
 * On macOS, use the more specific RemovePendingNotification or RemoveDeliveredNotification functions.
 *
 * @export
 * @param {string} identifier - The ID of the notification to remove
 * @return {Promise<void>}
 */
export function RemoveNotification(identifier) {
    return Call(":wails:RemoveNotification", [identifier]);
}

